services/src/machine_learning/xgb_config_builder.rs

```diff
@@ -0,0 +1,1057 @@
+                self.colsample_bynode.to_string(),
+            ),
+            ("lambda".to_owned(), self.lambda.to_string()),
+            ("alpha".to_owned(), self.alpha.to_string()),
+            ("tree_method".to_owned(), self.tree_method.to_string()),
+            ("sketch_eps".to_owned(), self.sketch_eps.to_string()),
+            (
+                "scale_pos_weight".to_owned(),
+                self.scale_pos_weight.to_string(),
+            ),
+            (
+                "refresh_leaf".to_owned(),
+                (u8::from(self.refresh_leaf)).to_string(),
+            ),
+            ("process_type".to_owned(), self.process_type.to_string()),
+            ("grow_policy".to_owned(), self.grow_policy.to_string()),
+            ("max_leaves".to_owned(), self.max_leaves.to_string()),
+            ("max_bin".to_owned(), self.max_bin.to_string()),
+            (
+                "num_parallel_tree".to_owned(),
+                self.num_parallel_tree.to_string(),
+            ),
+            ("predictor".to_owned(), self.predictor.to_string()),
+        ];
+
+        // Don't pass anything to XGBoost if the user didn't specify anything.
+        // This allows XGBoost to figure it out on it's own, and suppresses the
+        // warning message during training.
+        // See: https://github.com/davechallis/rust-xgboost/issues/7
+        if !self.updater.is_empty() {
+            v.push((
+                "updater".to_owned(),
+                self.updater
+                    .iter()
+                    .map(std::string::ToString::to_string)
+                    .collect::<Vec<String>>()
+                    .join(","),
+            ));
+        }
+
+        v
+    }
+}
+
+impl TreeBoosterParametersBuilder {
+    fn validate(&self) -> Result<(), String> {
+        Interval::new_closed_closed(0.0, 1.0).validate(&self.eta, "eta")?;
+        Interval::new_open_closed(0.0, 1.0).validate(&self.subsample, "subsample")?;
+        Interval::new_open_closed(0.0, 1.0).validate(&self.colsample_bytree, "colsample_bytree")?;
+        Interval::new_open_closed(0.0, 1.0)
+            .validate(&self.colsample_bylevel, "colsample_bylevel")?;
+        Interval::new_open_closed(0.0, 1.0).validate(&self.colsample_bynode, "colsample_bynode")?;
+        Interval::new_open_open(0.0, 1.0).validate(&self.sketch_eps, "sketch_eps")?;
+        Ok(())
+    }
+}
+
+impl From<String> for TreeMethod {
+    fn from(s: String) -> Self {
+        Self::from(s.borrow())
+    }
+}
+
+impl<'a> From<&'a str> for TreeMethod {
+    fn from(s: &'a str) -> Self {
+        match s {
+            "auto" => TreeMethod::Auto,
+            "exact" => TreeMethod::Exact,
+            "approx" => TreeMethod::Approx,
+            "hist" => TreeMethod::Hist,
+            "gpu_exact" => TreeMethod::GpuExact,
+            "gpu_hist" => TreeMethod::GpuHist,
+            // TODO: better error handling
+            _ => panic!("no known tree_method for {s}"),
```

Comment:
Implement TryFrom instead of From and return an error

Reply:
see: https://github.com/geo-engine/geoengine/pull/746#discussion_r1252302470